Mathlib/RingTheory/KrullDimension.lean

```diff
@@ -0,0 +1,123 @@
+theorem ringKrullDim_le_of_surjective (f : R →+* S) (hf : Function.Surjective f) :
+    ringKrullDim S ≤ ringKrullDim R :=
+  krullDim_le_of_strictMono (PrimeSpectrum.comap f) (Monotone.strictMono_of_injective
+    (fun _ _ hab ↦ Ideal.comap_mono hab) (PrimeSpectrum.comap_injective_of_surjective f hf))
+
+/-- If `I` is an ideal of `R`, then `ringKrullDim (R ⧸ I) ≤ ringKrullDim R`. -/
+theorem ringKrullDim_quotient_le (I : Ideal R) :
+    ringKrullDim (R ⧸ I) ≤ ringKrullDim R :=
+  ringKrullDim_le_of_surjective _ Ideal.Quotient.mk_surjective
+
+/-- If `R` and `S` are isomorphic, then `ringKrullDim R = ringKrullDim S`. -/
+theorem ringKrullDim_eq_of_ringEquiv (e : R ≃+* S) :
+    ringKrullDim R = ringKrullDim S :=
+  le_antisymm (ringKrullDim_le_of_surjective e.symm e.symm.surjective)
+    (ringKrullDim_le_of_surjective e e.surjective)
+
+/-- An alias for `ringKrullDim_eq_of_ringEquiv`. -/
+theorem _root_.RingEquiv.ringKrullDim (e : R ≃+* S) :
+    ringKrullDim R = ringKrullDim S :=
+  le_antisymm (ringKrullDim_le_of_surjective e.symm e.symm.surjective)
+    (ringKrullDim_le_of_surjective e e.surjective)
+
+section DimensionZero
+
+theorem _root_.PrimeSpectrum.asIdeal.isMaximal_of_ringKrullDim_eq_zero
+    (I : PrimeSpectrum R) (hdim : ringKrullDim R = 0) :
+    I.asIdeal.IsMaximal := by
+  have h := mem_maximals_of_krullDim_eq_zero hdim I
+  simp only [maximals, Set.top_eq_univ, Set.mem_univ, true_implies, true_and] at h
+  constructor; constructor
+  · exact Ideal.IsPrime.ne_top'
+  · intro J hIJ; by_contra hJ
+    obtain ⟨M, hM1, hM2⟩ := Ideal.exists_le_maximal J hJ
+    have := h $ show _ ≤ ⟨M, hM1.isPrime⟩ by exact (le_of_lt $ lt_of_lt_of_le hIJ hM2 : _ ≤ _)
+    exact (lt_self_iff_false _).1 $ lt_of_lt_of_le (lt_of_le_of_lt (by exact this) hIJ) hM2
+
+theorem _root_.Ideal.IsPrime.isMaximal_of_ringKrullDim_eq_zero
+    {I : Ideal R} (hI : I.IsPrime) (hdim : ringKrullDim R = 0) :
+    I.IsMaximal :=
+  PrimeSpectrum.asIdeal.isMaximal_of_ringKrullDim_eq_zero ⟨I, hI⟩ hdim
+
+theorem _root_.PrimeSpectrum.asIdeal.mem_minimalPrimes_of_ringKrullDim_eq_zero
+    (I : PrimeSpectrum R) (hdim : ringKrullDim R = 0) :
+    I.asIdeal ∈ minimalPrimes R := by
+  have h := mem_minimals_of_krullDim_eq_zero hdim I
+  simp only [minimals, Set.top_eq_univ, Set.mem_univ, true_implies, true_and] at h
+  exact ⟨⟨I.isPrime, OrderBot.bot_le I.asIdeal⟩, by
+    rintro J ⟨hJ, _⟩ hJI; exact h $ show ⟨J, hJ⟩ ≤ I by exact hJI⟩
+
+theorem _root_.Ideal.IsPrime.mem_minimalPrimes_of_ringKrullDim_eq_zero
+    {I : Ideal R} (hI : I.IsPrime) (hdim : ringKrullDim R = 0) :
+    I ∈ minimalPrimes R :=
+  PrimeSpectrum.asIdeal.mem_minimalPrimes_of_ringKrullDim_eq_zero ⟨I, hI⟩ hdim
+
+end DimensionZero
+
+section Field
+
+@[simp]
+theorem ringKrullDim_eq_zero_of_field (F : Type*) [Field F] : ringKrullDim F = 0 :=
+  krullDim_eq_zero_of_unique
+
+theorem ringKrullDim_eq_zero_of_isField {F : Type*} [CommRing F] (hF : IsField F) :
```

Comment:
Can we make this an iff for domains?

Reply:
I have formalised the theorem `ringKrullDim_eq_zero_iff_isField_of_isDomain`, which says that the Krull dimension of an integral domain is zero if and only if it is a field. Could you please help me check if any golfing can be done? Also, shall I preserve the theorems `ringKrullDim_eq_zero_of_field` and `ringKrullDim_eq_zero_of_isField` or delete them?

Reply:
> ringKrullDim_eq_zero_of_isField

But the statement of `ringKrullDim_eq_zero_of_isField` is different from that of `ringKrullDim_eq_zero_of_field`. One says that `isField` implies that the Krull dimension is 0; the other says that `Field` implies that the Krull dimension is 0. How can I make it an alias?

Reply:
> `ringKrullDim_eq_zero_of_field` should be kept. `ringKrullDim_eq_zero_of_isField` can be an `alias`. I was thinking along the lines of
> 
> ```
> theorem PrimeSpectrum.mem_minimalPrimes_iff {p : PrimeSpectrum R} :
>     p.asIdeal ∈ minimalPrimes R ↔ p ∈ minimals (· ≤ ·) Set.univ :=
>   have : asIdeal '' Set.univ = {p : Ideal R | p.IsPrime ∧ ⊥ ≤ p} := by
>     ext x; simpa using ⟨fun H ↦ H.choose_spec ▸ H.choose.2, fun H ↦ ⟨⟨_, H⟩, rfl⟩⟩
>   Iff.trans (this ▸ Iff.rfl) (map_mem_minimals_iff (s := LE.le) (f := asIdeal) (by simp) (by simp))
> 
> theorem ringKrullDim_eq_zero_iff_forall_mem_minimalPrimes [Nontrivial R] :
>   ringKrullDim R = 0 ↔ ∀ (I : Ideal R), I.IsPrime → I ∈ minimalPrimes R := by
>   refine krullDim_eq_zero_iff_mem_minimals_of_nonempty.trans ?_
>   simp_rw [← PrimeSpectrum.mem_minimalPrimes_iff]
>   exact ⟨(· ⟨·, ·⟩), (· _ <| PrimeSpectrum.isPrime ·)⟩
>   
> attribute [local instance] Ideal.bot_prime
> 
> theorem ringKrullDim_eq_zero_iff_isField_of_isDomain [IsDomain R] :
>     ringKrullDim R = 0 ↔ IsField R := by
>   simp only [ringKrullDim_eq_zero_iff_forall_mem_minimalPrimes, minimalPrimes,
>     Ideal.minimalPrimes_eq_subsingleton_self, Set.mem_singleton_iff,
>     ← not_iff_comm.mp Ring.not_isField_iff_exists_prime, ne_eq, not_exists, not_and, not_imp_not]
> ```
> 
> The first lemma is also useful for `Ideal.IsPrime.mem_minimalPrimes_of_ringKrullDim_eq_zero` and the second should be added anyways (also the maximal version).

But the statement of `ringKrullDim_eq_zero_of_isField` is different from that of `ringKrullDim_eq_zero_of_field`. One says that `isField` implies that the Krull dimension is 0; the other says that `Field` implies that the Krull dimension is 0. How can I make it an alias?

Reply:
I was originally thinking of `alias ⟨_, ringKrullDim_eq_zero_of_isField⟩ := ringKrullDim_eq_zero_iff_isField_of_isDomain` but you're right let's keep both.